src/monodroid/jni/embedded-assemblies.cc

```diff
@@ -119,120 +106,285 @@ EmbeddedAssemblies::install_preload_hooks ()
+	if (entry == nullptr) {
+		if (match->duplicate_count > 0 && match->duplicate_map != nullptr) {
+			log_debug (LOG_ASSEMBLY, "typemap: searching module [%s] duplicate map for token %u (0x%x)", mono_guid_to_string (mvid), token, token);
+			entry = binary_search <int32_t, TypeMapModuleEntry, compare_type_token> (&token, match->duplicate_map, match->duplicate_count);
+		}
 
-	if ((p->mapping == 0) ||
-			(p->num_entries <= 0) ||
-			(p->entry_length <= 0) ||
-			(p->value_offset >= p->entry_length) ||
-			(p->mapping == nullptr)) {
-		log_warn (LOG_DEFAULT, "Could not read type mapping file '%s!%s'. Ignoring...", source_apk, source_entry);
-		delete p;
-		return false;
+		if (entry == nullptr) {
+			log_warn (LOG_ASSEMBLY, "typemap: type with token %d (0x%x) in module {%s} (%s) not found.", token, token, mono_guid_to_string (mvid), match->assembly_name);
+			return nullptr;
+		}
 	}
 
-	p->source_apk   = strdup (source_apk);
-	p->source_entry = strdup (source_entry);
-	if (*info) {
-		(*info)->next = p;
+	uint32_t java_entry_count;
+#if defined (DEBUG) || !defined (ANDROID)
+	if (application_config.instant_run_enabled) {
+		java_entry_count = match->entry_count;
 	} else {
-		*info = p;
+#endif
+		java_entry_count = java_type_count;
+#if defined (DEBUG) || !defined (ANDROID)
 	}
-	return true;
+#endif
+	if (entry->java_map_index >= java_entry_count) {
+		log_warn (LOG_ASSEMBLY, "typemap: type with token %d (0x%x) in module {%s} (%s) has invalid Java type index %u", token, token, mono_guid_to_string (mvid), match->assembly_name, entry->java_map_index);
+		return nullptr;
+	}
+
+	const char *ret;
+#if defined (DEBUG) || !defined (ANDROID)
+	if (application_config.instant_run_enabled) {
+		ret = reinterpret_cast<char*>(match->java_map + ((match->java_name_width + 4) * entry->java_map_index));
+	} else {
+#endif
+		const TypeMapJava *java_entry = reinterpret_cast<const TypeMapJava*> (reinterpret_cast<const uint8_t*>(map_java) + ((sizeof(TypeMapJava) + java_name_width) * entry->java_map_index));
+		ret = reinterpret_cast<const char*>(reinterpret_cast<const uint8_t*>(java_entry) + 8);
+#if defined (DEBUG) || !defined (ANDROID)
+	}
+#endif
+
+	if (XA_UNLIKELY (ret == nullptr)) {
+		log_warn (LOG_ASSEMBLY, "typemap: empty Java type name returned for entry at index %u", entry->java_map_index);
+	}
+
+	if (XA_UNLIKELY (utils.should_log (LOG_TIMING))) {
+		total_time.mark_end ();
+
+		Timing::info (total_time, "Typemap.managed_to_java: end, total time");
+	}
+
+	log_debug (
+		LOG_ASSEMBLY,
+		"typemap: type with token %d (0x%x) in module {%s} (%s) corresponds to Java type '%s'",
+		token,
+		token,
+		mono_guid_to_string (mvid),
+		match->assembly_name,
+		ret
+	);
+
+	return ret;
+}
+
+int
+EmbeddedAssemblies::compare_type_token (const int32_t *token, const TypeMapModuleEntry *entry)
+{
+	if (entry == nullptr) {
+		log_warn (LOG_ASSEMBLY, "compare_type_token: entry is nullptr");
```

Comment:
Should this instead assert & crash the process?  Is there any plausible scenario for `entry` to be null?